Allow StringInput from borrowed str, to avoid copy

sunng87 opened the pull request:
~~Fixes 141~~ 

Added support for `StringInput` to hold a borrowed `str`, as well as `String`. This is to avoid String copy in functions like `parse_str`.

dragostis commented:
The problem with this is that it renders the `Rc` used everywhere in the API pretty much useless. (in `Position`, `Span`, `Pair` and `Pairs`) Maybe the right approach would be to replace the `Rc` with lifetimes? The only reason why `Rc`s were used in the first place was to keep the API simple.

sunng87 commented:
Yes, for correction, this patch doesn't fix issue mentioned in #141 . It's just to avoid allocating string when we use `parse_str`.

I had an attempt to replace `Rc<Input>` to `&'a Input`, it requires huge amount of keyboard work 😂 (I have got pest module compiles, but it still requires work for test and derive module)
For now, I'm OK with those Rc clones.

dragostis commented:
Then I propose a 2 part solution:

* implement `StrInput` which works with `&str`
* abstract over `StringInput` because they have the same implementation

After this is done, it may be worth researching whether replacing `Rc`s with borrows does not make the API extremely hard to use.

sunng87 commented:
Just added `StrInput` and removed unneeded lifetime indicators in `StringInput` and `FileInput`.

By the way, I'm using rustfmt with my editor so some files are formatted when I'm saving them. I recommend you to use rustfmt to keep the format consistent too. :-)

dragostis commented:
@sunng87 rustfmt is still not mature enough. I have a local `rustfmt.toml` file, but there are a few pending bugs in rustfmt that keep this from going anywhere.

dragostis commented on pest/src/inputs/string_input.rs:
```diff
@@ -143,7 +142,7 @@ unsafe fn line_of(source: &str, mut pos: usize) -> &str {
             |&(_,
```
This is one of the issues with rustfmt. 😛

dragostis commented on pest/src/inputs/string_input.rs:
```diff
@@ -30,10 +36,182 @@ impl StringInput {
+                pos -= c.len_utf8();
+                line_col = (line_col.0, line_col.1 + 1);
+            }
+            None => unreachable!(),
```
There should be no comma here. Hm, newest version ofrustfmt should have dealt with this particular case.

dragostis commented on pest/src/inputs/string_input.rs:
```diff
@@ -30,10 +36,182 @@ impl StringInput {
+
+        end
+    } else {
+        let end = source.char_indices().skip_while(|&(i, _)| i < pos).find(
```
This should be formatted like the one above.

sunng87 replied:
@dragostis I'll suggest you to put rustfmt.toml under the root directory of this project. So contributors get same rustfmt configuration and we can keep the code consistent. 

And for these formation issues, we can leave it for rustfmt iterations in future

dragostis replied:
I really considered that, but formatting can become really poor in some particular cases, like the one above, and it's not even consistent with the prior example. Putting the file there pretty much means keeping it. Right now, using rustfmt on the whole project makes the formatter hang because of an issues with nested lambdas. I am very much in favor of going down this road, but right now I would say it's partly too much of a hassle, and partly making the code messy like the example above.

sunng87 replied:
I see. I had issue with saving `pest/test/json.rs` and thought it was issue with Emacs' rust-mode.

dragostis commented on pest/src/inputs/span.rs:
```diff
@@ -182,7 +176,7 @@ impl<I: Input> PartialOrd for Span<I> {
             match self.start.partial_cmp(&other.start) {
                 Some(Ordering::Equal) => self.end.partial_cmp(&other.end),
-                ordering => ordering
+                ordering => ordering,
```
One little comma left. (updating rustfmt to latest master should fix it)

sunng87 commented:
I was using rustftm from crate.io, will update to master. 
Disabled format-on-save in emacs and manually formatted those lines.
